### PR TITLE
Implement finite-element stress solver

### PR DESCRIPTION
**Pull request by jlogan-cfs**
### Added

* Rust
    * Solenoid stress module with axisymmetric finite element system assembly
    * `mesh` submodules for quadrature rules and for each supported mesh and element type
    * Build python bindings with abi3-py310
    * Update rust dep versions
* Python
    * Bindings and examples for new axisymmetric FEM stress solver

### Changed

* Rust
    * Factor out quadrature rules and triangle mesh functions from BEM into mesh module
        * This is technically a breaking change, but documented here as a minor change because there are no direct users of the affected items yet
* Python
    * Fix docs build script

### CODE REVIEW

File: cfsem/solenoid_stress/axisymmetric_fem.py [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+    `element_type="quad4"` uses the input mesh directly. `element_type="quad9"` elevates the
+    corner-only input mesh to an explicit 9-node analysis mesh for the backend while keeping the
+    Python-side load and temperature inputs on the original corner nodes.
+    """
+
+    elements_arr = _normalize_elements(elements)
+    dtype = _resolve_float_dtype(nodes, material_table, thermal_material_table)
+    nodes_arr = _normalize_nodes(nodes, dtype)
+    material_ids_arr, material_table_arr = _normalize_materials(material_ids, material_table, dtype)
+    _thermal_ids_arr, thermal_material_table_arr = _normalize_thermal_material_table(
```
Comment (ian97531) [inline cfsem/solenoid_stress/axisymmetric_fem.py]: **Wrong-answer bug.** `_thermal_ids_arr` is computed against the thermal-table sort order and then discarded — only the elastic `material_ids_arr` (from `_normalize_materials` above) reaches the backend, which uses it to index the thermal table at [`loads/thermal.rs:137`](../blob/HEAD/src/physics/solenoid_stress/loads/thermal.rs#L137).

When `material_table` and `thermal_material_table` are both `Mapping`s with overlapping-but-different key sets, the two sort orders diverge and the backend silently looks up the wrong thermal row.

Concrete example producing wrong answers with no error:

- `material_ids = [10]`
- `material_table = {10: D10, 20: D20}` → sorted keys `[10, 20]` → `material_ids_arr = [0]`, dense `[D10, D20]`
- `thermal_material_table = {5: T5, 10: T10}` → sorted keys `[5, 10]` → dense `[T5, T10]`
- Backend does `thermal_table[0]` → `T5`, but the element's actual material is tagged `10`, which should pair with `T10`.

`require_mapping` only enforces mapping-vs-dense consistency, not key-set equality. The test suite doesn't catch this because all tests either use `material_ids=np.zeros(...)` or identical mapping keys.

**Fix:** either (a) pass a separate `thermal_material_ids_arr` through to the backend so the thermal lookup uses the thermal-table sort, or (b) require identical key sets between the two mappings and raise otherwise. (b) is the simpler invariant.

File: cfsem/solenoid_stress/axisymmetric_fem.py [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+        self,
+        nodal_temperature: ArrayLike | None,
+    ) -> npt.NDArray[np.floating[Any]] | None:
+        if self.n_temperature_nodes == 0:
+            values = (
+                np.zeros((0,), dtype=self.dtype)
+                if nodal_temperature is None
+                else np.asarray(nodal_temperature, dtype=self.dtype).reshape(-1)
+            )
+            assert values.size == 0, "nodal_temperature was provided, but this model has no thermal operator"
```
Comment (ian97531) [inline cfsem/solenoid_stress/axisymmetric_fem.py:369]: **Silent failure under `python -O`.** The `assert` is stripped when assertions are disabled, and the function then unconditionally returns `None`. The Rust re-check in `model.rs::normalize_temperature` only fires if Python sends `Some(non-empty)`, so it never runs — the user's `nodal_temperature` input is silently discarded on a model that has no thermal operator, and they'll attribute the zero thermal effect to something else.

Mirror the style of the `is required` branch just below:

```python
if values.size != 0:
    raise ValueError(
        "nodal_temperature was provided, but this model has no thermal operator"
    )
return None
```

File: cfsem/solenoid_stress/axisymmetric_fem.py [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+    else:
+        table = np.asarray(thermal_material_table, dtype=dtype)
+        assert (
+            table.ndim == 2 and table.shape[1] == 5
+        ), f"thermal_material_table must have shape (nmat, 5); got {table.shape}"
+        normalized_ids = np.ascontiguousarray(ids)
+        table = np.ascontiguousarray(table)
+    assert np.allclose(
+        table[:, 3], 0.0
+    ), "thermal_material_table shear thermal expansion must be zero in phase 1"
```
Comment (ian97531) [inline cfsem/solenoid_stress/axisymmetric_fem.py]: Three problems with these three lines:

1. **Strippable under `-O`.** `assert` disappears under `python -O`. The backend reads the full 4-component `alpha` vector via `constitutive_times_strain(material, &thermal.alpha)` at `loads/thermal.rs:45`, so a nonzero shear coefficient silently produces wrong thermal stress instead of failing.
2. **`np.allclose` default atol is `1e-8`.** A physically-negligible but numerically-nonzero `alpha_rz` slips past the check and then gets applied by the backend. Use exact comparison.
3. **"phase 1" has no referent.** Grep across the repo shows this is the only occurrence of the phrase — future maintainers have no way to recover what "phase 1" meant.

Suggested replacement:

```python
if np.any(table[:, 3] != 0.0):
    raise NotImplementedError(
        "shear thermal expansion (alpha_rz) is not yet supported"
    )
```

File: cfsem/solenoid_stress/axisymmetric_fem.py [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+    """Assemble the reusable axisymmetric FEM model.
+
+    The returned model stores the reduced stiffness matrix, the sparse load operators, the sparse
+    quadrature-recovery operators, and the fixed load topology associated with `pressure_faces`,
+    `traction_faces`, and `thermal_material_table`.
+
+    `element_type="quad4"` uses the input mesh directly. `element_type="quad9"` elevates the
+    corner-only input mesh to an explicit 9-node analysis mesh for the backend while keeping the
+    Python-side load and temperature inputs on the original corner nodes.
+    """
```
Comment (ian97531) [inline cfsem/solenoid_stress/axisymmetric_fem.py:957]: **Undocumented node-ordering convention.** The pressure-load outward-normal construction at `loads/pressure.rs:29` (`normal_area = [tangent[1], -tangent[0]]`) only yields outward normals when element nodes are numbered **counter-clockwise** in the `(r, z)` meridian plane. Clockwise elements are rejected downstream by the `det <= 0` guard in `src/mesh/elements/quad2d/mapping.rs` with the message `"non-positive element Jacobian determinant"` — which doesn't obviously point to the real issue for a user debugging mesh input.

Please add one sentence to this docstring (and the `AxisymmetricElements` enum doc on the Rust side) stating that `elements` must use CCW corner ordering in the `(r, z)` meridian plane. Also a good candidate for a dedicated validation error-path message.

File: cfsem/solenoid_stress/axisymmetric_fem.py [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+                continue
+            midpoint_index = int(elevated.analysis_elements[element_index, 4 + local_edge])
+            edge_to_midpoint[edge_key] = midpoint_index
+            rows.extend([midpoint_index, midpoint_index])
+            cols.extend([edge_key[0], edge_key[1]])
+            vals.extend([0.5, 0.5])
+        center_index = int(elevated.analysis_elements[element_index, 8])
+        rows.extend([center_index] * 4)
+        cols.extend([int(node) for node in conn])
+        vals.extend([0.25] * 4)
```
Comment (ian97531) [inline cfsem/solenoid_stress/axisymmetric_fem.py:660]: Nothing in the test suite pins these hand-rolled weights (`1.0` for corners, `0.5` for midsides, `0.25` for center) to analytical values. The uniform-temperature thermal test is linear-consistent-invariant, and the `np.linspace(nnode)` paths elsewhere only happen to be spatially-linear by accident of regular-mesh node ordering.

A refactor that changes `0.5` → `0.4` here would not fail CI.

Recommend adding a test that constructs `T = a*r + b*z` on the corner nodes, multiplies through this elevation operator, and asserts the elevated midside/center temperatures equal the analytic linear values exactly (linear interpolation must reproduce linear data).

File: cfsem/flux_solver.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    dx = xmesh[1][0] - xmesh[0][0]
+    dy = ymesh[0][1] - ymesh[0][0]
+
+    # Check regular grid assumption
+    assert np.all(
+        np.abs(np.diff(xmesh[:, 0]) - dx) / dx < 1e-6
+    ), "This method is only implemented for a regular grid"
+    assert np.all(
+        np.abs(np.diff(ymesh[0, :]) - dy) / dy < 1e-6
+    ), "This method is only implemented for a regular grid"
```
Comment (ian97531) [inline cfsem/flux_solver.py:77]: The regular-grid assertions at lines 72–77 are stripped under `python -O`, which silently lets this 4th-order stencil run on irregular grids and produce garbage gradients. Same pattern throughout this module: `_check_regular` (lines 275–286) positivity/monotonicity/regularity checks, and `calc_flux_density_from_flux`'s `rmesh > 0` check at line 132 — all cross-boundary numerical-correctness invariants using `assert`.

One sweep to convert these to `raise ValueError`.

File: cfsem/flux_solver.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        raise AssertionError(f"meshes and current_density must all have shape {expected_shape}")
+
+    # If the two axes have different lengths, the expected `indexing="ij"` layout
+    # is no longer ambiguous, so we can validate the mesh-axis content directly.
+    if rgrid.size != zgrid.size:
+        r_axis_matches = np.allclose(rmesh[:, 0], rgrid, rtol=tol, atol=tol)
+        z_axis_matches = np.allclose(zmesh[0, :], zgrid, rtol=tol, atol=tol)
+        assert (
+            r_axis_matches and z_axis_matches
+        ), "meshes must be consistent with grids and use np.meshgrid(..., indexing='ij')"
```
Comment (ian97531) [inline cfsem/flux_solver.py]: Two problems in `_validate_flux_mesh_inputs`:

1. **`raise AssertionError(...)` at line 196** is unusual — callers catching `ValueError` (the conventional bad-input exception) miss it. The preceding `assert not (...)` on line 190 is strip-eligible under `-O`, which degrades the helpful "appear transposed" diagnostic to the generic "must have shape" path.
2. **Square-grid bypass.** When `rgrid.size == zgrid.size`, the content-consistency check on lines 201–205 is skipped entirely. A user who accidentally passes a transposed meshgrid (a common mistake with `np.meshgrid` default `'xy'` indexing) silently gets wrong flux on every `nr == nz` mesh.

The docstring calls this "best-effort", which is honest but doesn't help the user. `rgrid.size == zgrid.size` only makes transpose-*detection* ambiguous; per-axis content equality (`rmesh[:, 0] == rgrid`, `zmesh[0, :] == zgrid`) is still checkable and should always run. Also consider converting both raises to `ValueError`.

File: cfsem/flux_solver.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    # Unpack and filter down to just useful inputs
+    dr, dz = _check_regular(grids, min_points=7)  # [m] grid spacing
+    area = dr * dz  # [m^2]
+    rmesh, zmesh = meshes  # [m]
+    assert not (
+        np.any(current_density[0, :] != 0.0)
+        or np.any(current_density[-1, :] != 0.0)
+        or np.any(current_density[:, 0] != 0.0)
+        or np.any(current_density[:, -1] != 0.0)
+    ), "current_density must be zero on the finite-difference boundary"
```
Comment (ian97531) [inline cfsem/flux_solver.py:245]: **Strippable cross-boundary invariant.** Under `python -O` the assert disappears. Nonzero boundary currents then flow into `ifil`/`rfil`/`zfil` at lines 261–263, and `flux_circular_filament` evaluates coincident filament/observation points on the boundary, which is singular — result is NaN flux values written back into `rhs` at line 265. Noisy failure under `-O` rather than a silently-wrong answer, but still an input-contract that should raise explicitly:

```python
if (np.any(current_density[0, :] != 0.0)
    or np.any(current_density[-1, :] != 0.0)
    or np.any(current_density[:, 0] != 0.0)
    or np.any(current_density[:, -1] != 0.0)):
    raise ValueError("current_density must be zero on the finite-difference boundary")
```

File: src/physics/solenoid_stress/types.rs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+#[derive(Clone, Copy, Debug)]
+pub struct PressureLoad<F: Real> {
+    /// Element index receiving the load.
+    pub element: usize,
+    /// Local face index in the element-family numbering used by `face_reference`.
+    pub local_face: u8,
+    /// Pressure magnitude, taken positive in the inward normal direction.
+    ///
+    /// Units: `[force / area]`.
+    pub value: F,
```
Comment (ian97531) [inline src/physics/solenoid_stress/types.rs]: **Dead field that implies behavior it doesn't have.** `PressureLoad::value` (and `TractionLoad::value` below on line 78) are never read by the load kernels — `pressure_face_kernel` and `traction_face_kernel` build unit-amplitude operators from `sample.tangent`/`sample.point`/`sample.n` only. Real amplitude flows in at `build_rhs` time via the `pressure_values` / `traction_values` arguments. The Python binding in `src/python.rs` hardcodes `value: F::one()` and `value: [F::zero(); 2]`.

A reader seeing this docstring will reasonably conclude that `PressureLoad { value: 5.0e6, .. }` applies 5 MPa. It does nothing.

Pick one:
- Delete the fields from both structs (and from the Python binding's `PressureLoad::new` / `TractionLoad::new`).
- Actually multiply by them in the kernels.
- Rewrite these docstrings to say "currently unused by assembly; amplitude is supplied at `build_rhs` time via the corresponding values array."

**Review by ian97531 — COMMENTED** [review]
These comments that I added are from Claude Code's review of the PR.

File: cfsem/solenoid_stress/axisymmetric_fem.py (hunk 2 above)
Comment (jlogan-cfs) [reply cfsem/solenoid_stress/axisymmetric_fem.py:369]: We do not support python -O

File: cfsem/solenoid_stress/axisymmetric_fem.py (hunk 1 above)
Comment (jlogan-cfs) [reply cfsem/solenoid_stress/axisymmetric_fem.py]: fixed, thanks!

File: cfsem/solenoid_stress/axisymmetric_fem.py (hunk 4 above)
Comment (jlogan-cfs) [reply cfsem/solenoid_stress/axisymmetric_fem.py:957]: fixed!

File: cfsem/solenoid_stress/axisymmetric_fem.py (hunk 3 above)
Comment (jlogan-cfs) [reply cfsem/solenoid_stress/axisymmetric_fem.py]: Fixed items 2 and 3; leaving item 1 because we do not support python -O

File: cfsem/solenoid_stress/axisymmetric_fem.py (hunk 5 above)
Comment (jlogan-cfs) [reply cfsem/solenoid_stress/axisymmetric_fem.py:660]: Added that test, thanks!

File: cfsem/flux_solver.py (hunk 6 above)
Comment (jlogan-cfs) [reply cfsem/flux_solver.py:77]: Not supporting python -O

File: cfsem/flux_solver.py (hunk 7 above)
Comment (jlogan-cfs) [reply cfsem/flux_solver.py]: Added the grid value comparison check, but continuing to not make changes to support python -O

File: cfsem/flux_solver.py (hunk 8 above)
Comment (jlogan-cfs) [reply cfsem/flux_solver.py:245]: not supporting python -O

File: src/physics/solenoid_stress/types.rs (hunk 9 above)
Comment (jlogan-cfs) [reply src/physics/solenoid_stress/types.rs]: Good find! Removed that, and the generic parameter that it used since that's not needed anymore

File: src/physics/solenoid_stress/recovery.rs [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+            for component in 0..4 {
+                let row = 4 * q_local + component;
+                let mut eps = 0.0;
+                let mut sig = 0.0;
+                for dof in 0..8 {
+                    eps += b[component][dof] * u[dof];
+                    sig += db[component][dof] * u[dof];
+                }
+                assert!((strain[row] - eps).abs() < 1.0e-12);
+                assert!((stress[row] - sig).abs() < 1.0e-3);
```
Comment (mdavids-cfs) [inline src/physics/solenoid_stress/recovery.rs:434]: Might be worth adding a comment here why you needed different tolerances

File: test/test_solenoid_stress_fem.py [hunk 11]
```diff
@@ -0,0 +1,1589 @@
...
+    samples = model.evaluate_quadrature(displacement, nodal_temperature=nodal_temperature)
+
+    assert np.allclose(samples.total_strain, samples.elastic_strain + samples.thermal_strain)
+
+
+@pytest.mark.parametrize("dtype", DTYPES, ids=lambda dtype: dtype.__name__)
+@pytest.mark.parametrize("quadrature", QUADRATURES)
+@pytest.mark.parametrize("element_type", ELEMENT_TYPES)
+@pytest.mark.parametrize("nr", PRESSURE_NR_CASES, ids=["coarse", "refined"])
+def test_pressure_vessel_stresses_match_lame_reference(
```
Comment (mdavids-cfs) [inline test/test_solenoid_stress_fem.py:1097]: Interesting proposal from Claude, that could help with the thermal model validation/coverage:
(see bottom "What's Missing" block)

 MMS = Method of Manufactured Solutions. You pick an exact displacement field
  u(r,z), substitute it into the governing PDE to derive the exact forcing (body
   force + boundary conditions + thermal strain) it would produce, run the
  solver with that forcing, and measure the error vs the known exact field as
  you refine the mesh. It's the gold-standard convergence test because:

  - It proves the code converges at the theoretically expected rate (quadratic
  for quad9, linear for quad4 in the energy norm) — not just that "finer →
  smaller residual."
  - It decouples correctness of assembly from correctness of solution method. A
  bug that halves the convergence rate (e.g., a wrong quadrature order, a bad
  B-matrix term for the hoop component) shows up immediately.
  - It catches sign/constant errors that uniform-field tests miss.

  What this PR already tests:
  - test_pressure_vessel_stresses_match_lame_reference — closed-form Lamé
  comparison (strong).
  - test_pressure_vessel_radial_displacement_matches_cfsem_1d_solver — grid
  refinement vs 1D solver, but tolerance checked at two mesh densities (nr=24,
  nr=48), not as an order-of-convergence fit.
  - Thermal:
  test_uniform_temperature_recovery_matches_fully_constrained_thermal_stress
  (test_solenoid_stress_fem.py:757) and its multi-material sibling (line 805) —
  both at constant temperature.
  - test_quad9_temperature_elevation_reproduces_affine_temperature_field — tests
   the quad4→quad9 interpolation only, not the thermal stress response to a
  spatially varying field.
  - The superposition test uses an affine nodal temperature T_ref + 18·rfrac +
  11·zfrac (line 894), but validates linearity, not correctness of the thermal
  coupling.

  What's missing:
  - A quadratic or non-polynomial nodal temperature field run through
  thermal_rhs + solve + recover, and compared against the analytical thermal
  stress distribution for a known case (e.g., radially-varying T(r) in a solid
  cylinder has a closed-form thermal stress).
  - An observed-rate fit: run at nr ∈ {8, 16, 32, 64}, log-log fit error vs h,
  assert slope ≥ expected order. This would lock in the quadratic behavior of
  quad9 and catch future regressions that silently drop the rate.

File: src/mesh/triangle3d.rs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        .chain(triangles.1.iter())
+        .chain(triangles.2.iter())
+        .any(|&idx| idx >= nnode)
+    {
+        return Err("Triangle refers to non-existent node");
+    }
+
+    for i in 0..ntri {
+        let idx = [triangles.0[i], triangles.1[i], triangles.2[i]];
+        if triangle_area_from_indices(nodes, idx) < 1e-14 {
```
Comment (mdavids-cfs) [inline src/mesh/triangle3d.rs]: Not sure if there is a good mechanism for throwing warnings in rust, but we could consider throwing warnings for near-sliver elements, i.e., testing against aspect ratios (1:10000 threshold or something similar) that would catastrophically
ill-condition the stiffness matrix. Not sure if that's super high priority, but might be useful down the road

File: src/mesh/quad2d.rs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        self.nodes_rz.len()
+    }
+
+    /// Number of elements.
+    pub fn num_elements(&self) -> usize {
+        self.elements.len()
+    }
+
+    /// Validate that every connectivity entry references an existing node.
+    pub fn validate_connectivity(&self) -> Result<(), String> {
```
Comment (mdavids-cfs) [inline src/mesh/quad2d.rs:26]: Could consider testing against duplicate nodes -- these will mechanically decouple the model locally; not sure that would be easy to spot in the solution, and not sure there would be cases where that would be intended; like jointly modeling plates that touch but are not connected?

**Review by mdavids-cfs — COMMENTED** [review]
Overall looking great! I had Claude run previously, and some/most of the thins it had found before are addressed. It had one interesting suggestion on testing the thermal pieces.
I also dug into the mesh piece a little bit, and found two tests we could add, aspect ratio is likely optional, the duplicate node one we may want to add.

File: test/test_solenoid_stress_fem.py (hunk 11 above)
Comment (jlogan-cfs) [reply test/test_solenoid_stress_fem.py:1097]: Good idea! Added a convergence rate assert in the convergence example and a comparison to 1D thermal stress handcalc.

File: src/mesh/quad2d.rs (hunk 13 above)
Comment (jlogan-cfs) [reply src/mesh/quad2d.rs:26]: Yeah, there are cases where we want duplicate nodes for modeling purposes, but we could give a warning when they're present

Comment (jlogan-cfs) [reply src/mesh/quad2d.rs:26]: Added a warning on duplicate points

File: src/mesh/triangle3d.rs (hunk 12 above)
Comment (jlogan-cfs) [reply src/mesh/triangle3d.rs]: Good idea, I'll add that as well

Comment (jlogan-cfs) [reply src/mesh/triangle3d.rs]: Added a warning on low quality triangles

File: src/physics/solenoid_stress/recovery.rs (hunk 10 above)
Comment (jlogan-cfs) [reply src/physics/solenoid_stress/recovery.rs:434]: Added a note about stress recovery going through the gradient of the strain, which introduces error

**Review by mdavids-cfs — APPROVED** [review]
Looking great! \o/
